library/asn1parse.c

```diff
@@ -40,7 +40,7 @@ int mbedtls_asn1_get_len(unsigned char **p,
                          const unsigned char *end,
                          size_t *len)
 {
-    if ((end - *p) < 1) {
+    if (end <= *p) {
```

Comment:
`if (end <= *p)` is clearer than `if ((end - *p) < 1)` in isolation. But those lines are not in isolation: they're part of an idiom throughout the ASN.1 and X.509 parsing code that uses `if (end - *p <= n)` as a check before reading `n` bytes. I think the code base should keep this uniform idiom, rather than simplify individual cases.

Reply:
The reason for this change is actually robustness.  If Mbed TLS is compiled with `-fno-strict-overflow` (which it probably should be, unless `-fsanitize=undefined -fsanitize-minimal-runtime` is an option), then `end <= *p` might work in cases where `(end - *p) < 1` fails due to overflow, specifically when `end` and `*p` differ by half of the address space or more.

Reply:
That's a good argument, but I'm convinced it's a good way to do it. How about changing the code to
```
if (end < *p) {
    /* Indicates a bug in user code. Error out explicitly for robustness. */
    return MBEDTLS_ERR_ASN1_OUT_OF_DATA;
}
if (end <= *p) {
    return MBEDTLS_ERR_ASN1_OUT_OF_DATA;
}
```

Or, for that matter, since the end-already-reached case is an application error rather than bad input data, maybe the code should return a different error? (Which will cost a couple of bytes of code size, but this is only useful in `get_len` and `get_tag`, since other functions call one of these first.)
```
if (end < *p) {
    /* Indicates a bug in user code. Error out explicitly for robustness. */
    return MBEDTLS_ERR_CORRUPTION_DETECTED;
}
if (end <= *p) {
    return MBEDTLS_ERR_ASN1_OUT_OF_DATA;
}
```

Mind you, _in practice_,

Reply:
I’m not sure, but I can find out when I have some free time.